### PR TITLE
[srp-client] add API to specify TTL

### PR DESCRIPTION
**Pull request by jwhui**
---
The SRP Server and Advertising Proxy also need updating to make use of the TTL.

### CODE REVIEW

**Review by abtink — APPROVED** [review]
Looks great. Thanks. 👍

File: include/openthread/srp_client.h [hunk 1]
```diff
@@ -315,6 +315,30 @@ void otSrpClientDisableAutoStartMode(otInstance *aInstance);
  */
 bool otSrpClientIsAutoStartModeEnabled(otInstance *aInstance);
 
+/**
+ * This function gets the TTL used in SRP update requests.
```
Comment (tcarmelveilleux) [inline include/openthread/srp_client.h:319]: Suggest mentioning every record has the same TTL in SRP server

File: src/core/net/srp_client.cpp [hunk 2]
```diff
@@ -1137,7 +1138,7 @@ Error Client::AppendUpdateLeaseOptRecord(Message &aMessage) const
     }
     else
     {
-        leaseOption.SetLeaseInterval(mLeaseInterval);
+        leaseOption.SetLeaseInterval(GetTtl());
```
Comment (tcarmelveilleux) [inline src/core/net/srp_client.cpp:1141]: Curious why this change here. Should this not remain based on lease time?

File: src/core/net/srp_client.hpp [hunk 3]
```diff
@@ -410,6 +410,27 @@ class Client : public InstanceLocator, private NonCopyable
      */
     void SetCallback(Callback aCallback, void *aContext);
 
+    /**
+     * This method gets the TTL used in SRP update requests.
+     *
+     * Note that this is the TTL that would be requested by the SRP client. Server may choose to accept a different TTL.
+     *
+     * @returns The TTL (in seconds).
```
Comment (tcarmelveilleux) [inline src/core/net/srp_client.hpp:418]: Suggest mentioning that value 0 is a special case and causes TTL to follow lease time

File: src/core/net/srp_client.cpp (hunk 2 above)
Comment (abtink) [reply src/core/net/srp_client.cpp:1141]: I agree. I think we may want to stay `mLeaseInterval`.

File: src/core/net/srp_client.hpp (hunk 3 above)
Comment (jwhui) [reply src/core/net/srp_client.hpp:418]: This will always return the TTL value used in SRP Update requests. If `mTtl` is set to zero, this will return `mLeaseInterval`.

File: src/core/net/srp_client.hpp [hunk 4]
```diff
@@ -410,6 +410,27 @@ class Client : public InstanceLocator, private NonCopyable
...
+     *
+     */
+    uint32_t GetTtl(void) const { return mTtl ? mTtl : mLeaseInterval; }
```
Comment (abtink) [inline src/core/net/srp_client.hpp:421]: What if `mLeaseInterval` is shorter than specified `mTtl` ? 
Does it make sense to do something like `mTtl ? OT_MIN(mTtl, mLeaseInterval) : mLeaseInterval`?

Comment (abtink) [reply src/core/net/srp_client.hpp:421]: I checked the SRP draft. I think we want to use the min.

>    TTLs sent in SRP Updates are advisory: they indicate the SRP client's
>    guess as to what a good TTL would be.  SRP servers may override these
>    TTLs.  SRP servers SHOULD ensure that TTLs are reasonable: neither
>    too long nor too short.  **The TTL should never be longer than the
>    lease time** .

Comment (jwhui) [reply src/core/net/srp_client.hpp:421]: Things get a bit more complicated when the user changes the lease interval too. I'm inclined to let the server handle it. - see https://datatracker.ietf.org/doc/html/draft-ietf-dnssd-srp#section-3. Thoughts?

Comment (jwhui) [reply src/core/net/srp_client.hpp:421]:
```
TTLs sent in SRP Updates are advisory: they indicate the SRP client's
guess as to what a good TTL would be. SRP servers may override these
TTLs. SRP servers SHOULD ensure that TTLs are reasonable: neither
too long nor too short. The TTL should never be longer than the
lease time .
```

I read this as the SRP server should be responsible for overriding the requested TTL value. So should we keep the client simple and let the server handle it?

Comment (abtink) [reply src/core/net/srp_client.hpp:421]: I agree. The server has the authority to actually accept the lease interval. The lease from clinet is again its desire.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/core/net/srp_client.hpp [hunk 5]
```diff
@@ -410,6 +410,27 @@ class Client : public InstanceLocator, private NonCopyable
...
+    /**
+     * This method sets the TTL used in SRP update requests.
+     *
+     * Changing the TTL does not impact the TTL of already registered services/host-info.
+     * It only changes any future SRP update messages (i.e adding new services and/or refreshes of existing services).
+     *
+     * @param[in] aTtl  The TTL (in seconds). If zero, the TTL is set to the lease interval.
+     *
+     */
+    void SetTtl(uint32_t aTtl) { mTtl = aTtl; }
```
Comment (simonlingoogle) [inline src/core/net/srp_client.hpp:432]: What's the relationship between TTL and Lease Time?

Should the TTL always be smaller than Lease Time?

**Review by simonlingoogle — COMMENTED** [review]
Why is TTL not configured per host or service ?

File: src/core/net/srp_client.hpp (hunk 4 above)
Comment (simonlingoogle) [reply src/core/net/srp_client.hpp:421]: I am inclined to think that we should not send TTL larger than Lease Time in the first place, or even force it into the SRP client implementation. For example, we should not allow setting TTL that is larger than Lease Time. 

Thoughts?

Comment (jwhui) [reply src/core/net/srp_client.hpp:421]: Both TTL and lease time in an SRP Update Request are proposals. Ultimately, the server gets to choose the values. If the server chooses a smaller lease time, then the server also needs to choose a smaller TTL. As a result, the SRP Server already needs to incorporate the logic of setting the TTL appropriately. It's not clear to me that there is much value in enforcing it on the client side.

File: src/core/net/srp_client.hpp (hunk 5 above)
Comment (wgtdkp) [reply src/core/net/srp_client.hpp:432]: The lease time controls when the SRP server deletes an outdated SRP registration, while TTL remains the same meaning as in normal DNS records. The SRP spec may help: https://datatracker.ietf.org/doc/html/draft-ietf-dnssd-srp-13#section-3

**Review by wgtdkp — APPROVED** [review]
LGTM 👍

File: src/core/net/srp_client.hpp (hunk 5 above)
Comment (jwhui) [reply src/core/net/srp_client.hpp:432]: Current draft spec states that TTL should never be longer than the lease time, but not a must:

```
   TTLs sent in SRP Updates are advisory: they indicate the SRP client's
   guess as to what a good TTL would be.  SRP servers may override these
   TTLs.  SRP servers SHOULD ensure that TTLs are reasonable: neither
   too long nor too short.  The TTL should never be longer than the
   lease time (Section 4.1).  Shorter TTLs will result in more frequent
   data refreshes; this increases latency on the DNS-SD client side,
   increases load on any caching resolvers and on the authoritative
   server, and also increases network load, which may be an issue for
   constrained networks.  Longer TTLs will increase the likelihood that
   data in caches will be stale.  TTL minimums and maximums SHOULD be
   configurable by the operator of the SRP server.
```

Comment (simonlingoogle) [reply src/core/net/srp_client.hpp:432]: Should we stick to `SHOULD` requirement if there is no strong reason not to?

Comment (jwhui) [reply src/core/net/srp_client.hpp:432]: The SHOULD in the text above applies to servers. The client values are advisory. The server is the one that technically assigns the value that is used.

Comment (jwhui) [reply src/core/net/srp_client.hpp:436]: Regardless, I updated `GetTtl()` and relevant docs to indicate that lease interval is used if TTL is configured to a larger value.

**Review by simonlingoogle — APPROVED** [review]
LGTM 👍
